src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g

```diff
@@ -1390,6 +1396,15 @@ relationalExpression
         )
     ;
 
+instanceofTypeSpec!
+    :   t:typeSpec[true]
+        {#instanceofTypeSpec = #t;}
+    ;
+
+patternDefinition!
+    :       t:typeSpec[false] v:variableDefinitions[#null,#t]
+            {#patternDefinition = #(#[PATTERN_DEF,"PATTERN_DEF"], v);}
```

Comment:
I am questioning new token creation .....


`if (o instanceof String s) {` is like `<variable name> instanceof <variable declaration>`
 
```
$ cat Test.java 
public class Test {
 String field;
}
$ java $RUN_LOCALE -jar /var/tmp/checkstyle-8.33-all.jar -t Test.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> Test [1:13]
`--OBJBLOCK -> OBJBLOCK [1:18]
    |--LCURLY -> { [1:18]
    |--VARIABLE_DEF -> VARIABLE_DEF [2:1]
    |   |--MODIFIERS -> MODIFIERS [2:1]
    |   |--TYPE -> TYPE [2:1]
    |   |   `--IDENT -> String [2:1]
    |   |--IDENT -> field [2:8]
    |   `--SEMI -> ; [2:13]
    `--RCURLY -> } [3:0]
```

Why we can not reuse `VARIABLE_DEF` ?

All Checks that use VARIABLE_DEF will be affected, and this is good, as without this users can put any name to it and checkstyle will not catch it.

Please create Input update for VariableName Check.

Reply:
>I am questioning new token creation .....

So, this:
```
➜  initial-pattern-def-grammar cat TestClass.java 
public class TestClass {
 {
	Object o;
	if (o instanceof String s) { 
	}
 }
}

```
Should be this
```
➜  initial-pattern-def-grammar java -jar ~/IdeaProjects/checkstyle/target/checkstyle-8.34-SNAPSHOT-all.jar -T TestClass.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> TestClass [1:13]
...
    |       |--LITERAL_IF -> if [4:1]
    |       |   |--LPAREN -> ( [4:4]
    |       |   |--EXPR -> EXPR [4:7]
    |       |   |   `--LITERAL_INSTANCEOF -> instanceof [4:7]
    |       |   |       |--IDENT -> o [4:5]
    |       |   |       `--VARIABLE_DEF -> VARIABLE_DEF [4:18]
    |       |   |           |--TYPE -> TYPE [4:18]
    |       |   |           |   `--IDENT -> String [4:18]
    |       |   |           `--IDENT -> s [4:25]
    |       |   |--RPAREN -> ) [4:26]
...

```
not this?
```
➜  initial-pattern-def-grammar java -jar ~/IdeaProjects/checkstyle/target/checkstyle-8.34-SNAPSHOT-all.jar -T TestClass.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> TestClass [1:13]
...
    |       |--LITERAL_IF -> if [4:1]
    |       |   |--LPAREN -> ( [4:4]
    |       |   |--EXPR -> EXPR [4:7]
    |       |   |   `--LITERAL_INSTANCEOF -> instanceof [4:7]
    |       |   |       |--IDENT -> o [4:5]
    |       |   |       `--PATTERN_DEF -> PATTERN_DEF [4:18]
    |       |   |           `--VARIABLE_DEF -> VARIABLE_DEF [4:18]
    |       |   |               |--TYPE -> TYPE [4:18]
    |       |   |               |   `--IDENT -> String [4:18]
    |       |   |               `--IDENT -> s [4:25]
    |       |   |--RPAREN -> ) [4:26]
...
``` ?
>Please create Input update for LocaVariableName Check and in all other Checks that should care about variables

Should I create new files, or add to the existing ones?

Reply:
> I am questioning new token creation .....
From https://openjdk.java.net/jeps/305
```
RelationalExpression:
     ...
     RelationalExpression instanceof ReferenceType
     RelationalExpression instanceof Pattern

Pattern:
     ReferenceType Identifier
```

As of today, the pattern looks like a variable declaration. But this may be extended in a future version of Java.
Here is a good overview of pattern matching: https://www.youtube.com/watch?v=qurG_J81_Cs&t=982

Reply:
Good video, to view that there will probably be a lot of ideas and extensions and .... .
lets do:
```
    |       |   |   `--LITERAL_INSTANCEOF -> instanceof [4:7]
    |       |   |       |--IDENT -> o [4:5]
    |       |   |       `--PATTERN_DEF -> PATTERN_DEF [4:18]
    |       |   |           `--VARIABLE_DEF -> VARIABLE_DEF [4:18]
    |       |   |               |--TYPE -> TYPE [4:18]
    |       |   |               |   `--IDENT -> String [4:18]
    |       |   |               `--IDENT -> s [4:25]
```

Reply:
Done.